### PR TITLE
feat(`foundry-common`): NameOrAddress ENS util

### PR DESCRIPTION
**Pull request by Evalir**
## Motivation

We need a foundry-specific utility to resolve ens names at the CLI level.

## Solution

Adds `NameOrAddress` for CLI use, which can resolve itself to an address, or error.

Part of #7106

### CODE REVIEW

File: crates/common/src/ens.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    Name(String),
+    /// An Ethereum Address
+    Address(Address),
+}
+
+impl NameOrAddress {
+    /// Resolves the name to an Ethereum Address.
+    pub async fn resolve<P: TempProvider>(
+        &self,
+        provider: Arc<P>,
```
Comment (DaniPopes) [inline crates/common/src/ens.rs:42]:
```suggestion
        provider: &P,
```

File: crates/common/src/ens.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+pub enum NameOrAddress {
+    /// An ENS Name (format does not get checked)
+    Name(String),
+    /// An Ethereum Address
+    Address(Address),
+}
+
+impl NameOrAddress {
+    /// Resolves the name to an Ethereum Address.
+    pub async fn resolve<P: TempProvider>(
```
Comment (DaniPopes) [inline crates/common/src/ens.rs:40]: should this be an extension crate in alloy?

File: crates/common/src/ens.rs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    }
+}
+
+impl From<Address> for NameOrAddress {
+    fn from(addr: Address) -> Self {
+        NameOrAddress::Address(addr)
+    }
+}
+
+impl From<&str> for NameOrAddress {
```
Comment (DaniPopes) [inline crates/common/src/ens.rs:85]: not a fan of different behavior for different string types

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/common/src/ens.rs (hunk 2 above)
Comment (Evalir) [reply crates/common/src/ens.rs:40]: Yeah I think we can eventually move this to alloy—I'd like to do this once the provider's a bit more stable though, after this PR

File: crates/common/src/ens.rs [hunk 4]
```diff
@@ -0,0 +1,143 @@
+//! ENS Name resolving utilities.
+#![allow(missing_docs)]
+use std::str::FromStr;
+
```
Comment (mattsse) [inline crates/common/src/ens.rs:4]:
```suggestion
```

File: crates/common/src/ens.rs (hunk 2 above)
Comment (mattsse) [reply crates/common/src/ens.rs:40]: This type should be in alloy yes

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/common/src/ens.rs (hunk 2 above)
Comment (Evalir) [reply crates/common/src/ens.rs:39]: let's do it after https://github.com/alloy-rs/alloy/pull/190 is merged, else we'll need to refactor twice

**Review by mattsse — APPROVED** [review]

**Pull request merged**